### PR TITLE
Key assignment type safety

### PR DESCRIPTION
**Pull request by smarshall-spitzbart**
# Description

Reduces foot guns in key assignment code which previously used `sdk.ConsAddress` for both provider and consumer addresses. This PR makes key assignment code leverage type safety better by introducing protobuf generated wrappers around consensus addresses, `ConsumerConsAddress` and `ProviderConsAddress`

Note that key assignment state is still persisted as hand crafted `sdk.ConsAddress` bytes, but future refactors could use these new pb types to store key assignment state in KV store value bytes using pb serialization methods.

## Linked issues

Closes: https://github.com/cosmos/interchain-security/issues/707

Note this PR enables a good solution to #728, but I'll save that as future work.

## Type of change

- [x] `Refactor`: Changes existing code style, naming, structure, etc.

## Regression tests

Existing key assignment tests should suffice, lmk if there's more targeted tests I should develop.

### CODE REVIEW

File: x/ccv/provider/keeper/key_assignment.go [hunk 1]
```diff
@@ -566,6 +566,8 @@ func (k Keeper) MustApplyKeyAssignmentToValUpdates(
 
 // GetProviderAddrFromConsumerAddr returns the consensus address of a validator with
 // consAddr set as the consensus address on a consumer chain
+//
+// TODO: communicate better that this method may return a consumer addr
```
Comment (glnro) [inline x/ccv/provider/keeper/key_assignment.go:570]: What would precipitate this function returning a consumer address?

Comment (MSalopek) [reply x/ccv/provider/keeper/key_assignment.go:570]:
```golang
	// check if this address is known only to the consumer chain
	if providerConsAddr, found := k.GetValidatorByConsumerAddr(ctx, chainID, consAddr); found {
		return providerConsAddr
	}
	return consAddr
```

As it stands now, it is a bit messy.

Consumer address is equal to provider address in case the operator does not use key assignment (ie, they use the same validator key for provider and consumer)

File: proto/interchain_security/ccv/provider/v1/provider.proto [hunk 2]
```diff
@@ -196,9 +191,49 @@ message VscSendTimestamp {
   google.protobuf.Timestamp timestamp = 2 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
 }
 
+//
+// Key assignment section
+//
+
+message ConsumerConsAddress {
```
Comment (MSalopek) [inline proto/interchain_security/ccv/provider/v1/provider.proto:198]: An edgecase crossed my mind:
What if the address format is different on provider and consumer?

eg. `cosmosvalconsjdfp0gks0dr9gs` vs `stridevalconsjdfp0gks0dr9gs`

Do we have/need any special handling in that case?

**Comment by smarshall-spitzbart** [comment]
Totally my bad, this is very much a draft PR lol

File: x/ccv/provider/keeper/key_assignment.go (hunk 1 above)
Comment (smarshall-spitzbart) [reply x/ccv/provider/keeper/key_assignment.go:570]: @MSalopek has the correct understanding. That code behavior was not clear to me from reading code/comments

File: proto/interchain_security/ccv/provider/v1/provider.proto (hunk 2 above)
Comment (smarshall-spitzbart) [reply proto/interchain_security/ccv/provider/v1/provider.proto:198]: I don't believe this is an issue, but I'm not 100% sure. These are consensus addresses that are persisted as bytes. `ConsumerCoonsAddress` is just a type wrapper for a byte array

File: x/ccv/provider/keeper/genesis.go [hunk 3]
```diff
@@ -73,16 +73,16 @@ func (k Keeper) InitGenesis(ctx sdk.Context, genState *types.GenesisState) {
 
 	// Import key assignment state
 	for _, item := range genState.ValidatorConsumerPubkeys {
-		k.SetValidatorConsumerPubKey(ctx, item.ChainId, item.ProviderAddr, *item.ConsumerKey)
+		k.SetValidatorConsumerPubKey(ctx, item.ChainId, *item.ProviderAddr, *item.ConsumerKey)
```
Comment (smarshall-spitzbart) [inline x/ccv/provider/keeper/genesis.go:76]: I believe these changes mean v1.1 will require state migrations from v1.0

File: x/ccv/provider/keeper/hooks_test.go [hunk 4]
```diff
@@ -33,7 +33,10 @@ func TestValidatorConsensusKeyInUse(t *testing.T) {
 			setup: func(ctx sdk.Context, k providerkeeper.Keeper) {
```
Comment (smarshall-spitzbart) [inline x/ccv/provider/keeper/hooks_test.go:33]: This test was hard to follow in some cases but I did not change which validator addrs are input to the methods

File: x/ccv/provider/keeper/key_assignment.go [hunk 5]
```diff
@@ -128,12 +130,16 @@ func (k Keeper) GetValidatorByConsumerAddr(
...
-	consumerAddr sdk.ConsAddress,
-	providerAddr sdk.ConsAddress,
+	consumerAddr types.ConsumerConsAddress,
+	providerAddr types.ProviderConsAddress,
 ) {
 	store := ctx.KVStore(k.storeKey)
 	// Cons address is a type alias for a byte string, no marshaling needed
-	bz := providerAddr
+	bz, err := providerAddr.Marshal()
+	if err != nil {
```
Comment (smarshall-spitzbart) [inline x/ccv/provider/keeper/key_assignment.go:139]: Technically a change in behavior, however an invalid cons address getting persisted would eventually cause critical errors. Addresses should be validated by this point (correct me if I'm wrong)

See `TestValidatorByConsumerAddrCRUD`

File: x/ccv/provider/keeper/key_assignment_test.go [hunk 6]
```diff
@@ -873,18 +878,18 @@ func TestSimulatedAssignmentsAndUpdateApplication(t *testing.T) {
 			// Build up the historicSlashQueries data structure
 			for i := range consumerValset.identities {
 				// For each active validator on the consumer chain
-				consC := consumerValset.identities[i].SDKValConsAddress()
+				consC := consumerValset.identities[i].ConsumerConsAddress()
 				if 0 < consumerValset.power[i] {
 					// Get the provider who assigned the key
 					consP := k.GetProviderAddrFromConsumerAddr(ctx, CHAINID, consC)
 
-					if _, found := historicSlashQueries[string(consC)]; !found {
```
Comment (smarshall-spitzbart) [inline x/ccv/provider/keeper/key_assignment_test.go:881]: This test was hard to follow but I did my best to keep its intention the same

File: x/ccv/provider/keeper/throttle.go [hunk 7]
```diff
@@ -195,7 +195,12 @@ func (k Keeper) GetSlashMeterAllowance(ctx sdktypes.Context) sdktypes.Int {
 func (k Keeper) QueueGlobalSlashEntry(ctx sdktypes.Context, entry providertypes.GlobalSlashEntry) {
 	store := ctx.KVStore(k.storeKey)
 	key := providertypes.GlobalSlashEntryKey(entry)
-	store.Set(key, entry.ProviderValConsAddr)
+	bz, err := entry.ProviderValConsAddr.Marshal()
+	if err != nil {
+		// This should never happen, since the provider val cons addr should be a valid sdk address
```
Comment (smarshall-spitzbart) [inline x/ccv/provider/keeper/throttle.go:200]: We again assume the provider cons address has been properly validated by now. This is where it's obtained https://github.com/cosmos/interchain-security/blob/38a1540972d3cd1dda7285c5a602b8706840270f/x/ccv/provider/keeper/relay.go#L316

File: x/ccv/provider/types/keys.go [hunk 8]
```diff
@@ -262,20 +262,20 @@ func ParseVscSendingTimestampKey(bz []byte) (string, uint64, error) {
 
 // ConsumerValidatorsKey returns the key under which the
 // validator assigned keys for every consumer chain are stored
-func ConsumerValidatorsKey(chainID string, addr sdk.ConsAddress) []byte {
-	return ChainIdAndConsAddrKey(ConsumerValidatorsBytePrefix, chainID, addr)
+func ConsumerValidatorsKey(chainID string, addr ProviderConsAddress) []byte {
```
Comment (smarshall-spitzbart) [inline x/ccv/provider/types/keys.go:265]: For these methods, provider vs consumer cons addr type was not previously specified. However I inspected code and found that all callers gave one of two types

**Comment by smarshall-spitzbart** [comment]
Full `make test` passed locally for me on latest commit

File: tests/e2e/slashing.go [hunk 9]
```diff
@@ -147,7 +148,7 @@ func (s *CCVTestSuite) TestRelayAndApplyDowntimePacket() {
 	// check that the outstanding slashing flag is reset on first consumer,
 	// since that consumer originally sent the slash packet.
 	// It's important to use the consumer's cons addr here.
-	pFlag := firstConsumerKeeper.OutstandingDowntime(s.consumerCtx(), consumerConsAddr)
+	pFlag := firstConsumerKeeper.OutstandingDowntime(s.consumerCtx(), consumerConsAddr.ToSdkConsAddr())
```
Comment (sainoe) [inline tests/e2e/slashing.go:151]: Perhaps the `OutstandingDowntime` keeper's functions should also take a `ConsumerConsAddress` rather that `sdk.ConsAddress`.

File: proto/interchain_security/ccv/provider/v1/provider.proto [hunk 10]
```diff
@@ -196,9 +191,57 @@ message VscSendTimestamp {
...
+}
+
+// AddressList contains a list of consumer consensus addresses
+message ConsumerAddressList {
+  repeated ConsumerConsAddress addresses = 1;
+}
+  
 message KeyAssignmentReplacement {
-  bytes provider_addr = 1;
+  ProviderConsAddress provider_addr = 1;
```
Comment (sainoe) [inline proto/interchain_security/ccv/provider/v1/provider.proto:220]: The CLI fills the `KeyAssignmentReplacement` msg using the validator operator address
see how we it's parsed [here](https://github.com/cosmos/interchain-security/blob/bf6cdae5f02ff31bb2982766bed9db4c664a0ec3/x/ccv/provider/keeper/msg_server.go#L38)

File: proto/interchain_security/ccv/provider/v1/provider.proto [hunk 11]
```diff
@@ -196,9 +191,57 @@ message VscSendTimestamp {
...
+}
+
+// A validator's consensus address on the provider chain
+message ProviderConsAddress {
+  // Do not generate stringer for this type, we'll use sdk.ConsAddress.String() instead
+  option (gogoproto.goproto_stringer) = false;
+  bytes address = 1;
+}
+
+// AddressList contains a list of consumer consensus addresses
```
Comment (sainoe) [inline proto/interchain_security/ccv/provider/v1/provider.proto:214]: nit: //ConsumerAddressList

**Comment by sainoe** [comment]
Great work @smarshall-spitzbart! LGTM. 
As a suggestion, we could use a `Bytes() ` method returning the raw address bytes of the `ConsumerConsAddress` and `ProviderConsAddress` in the serialization rather than using `ToSdkConsAddress()`.

**Review by sainoe — APPROVED** [review]

**Comment by smarshall-spitzbart** [comment]
> Great work @smarshall-spitzbart! LGTM. As a suggestion, we could use a `Bytes() ` method returning the raw address bytes of the `ConsumerConsAddress` and `ProviderConsAddress` in the serialization rather than using `ToSdkConsAddress()`.

Thanks for the suggestion! I went with the current design just because there were some spots in code that previously accepted the `sdk.ConsAddress` type, and the `ToSdkConsAddress()` helper allows the changes to be more of a drop-in replacement. `sdk.ConsAddress` is still automatically casted to byte array where that type is accepted

**Review by MSalopek — APPROVED** [review]
Awesome job!

Full integration tests passed on the pipeline.
